Remove duplicate live_offset entries from frametables

kayceesrk opened the pull request:
`Reg.Set` uses stamp for comparison. Hence, two virtual registers that map to the same real register may have different stamps. This leads to duplicate live_offset entries in the frametables. For example, on amd64,

```
$ wget https://raw.githubusercontent.com/dbuenzli/nbcodec/master/src/se.ml
$ ocamlopt.opt -S se.ml
```

See https://gist.github.com/kayceesrk/82c88ac67dcd7281a27a#file-se-s-L4523-L4528 for the output. The entry corresponds to the frametable for label .L375. The entry says that there are 3 live registers with offsets 3,3,5 which correspond to %rbx, %rbx, %rdi. The duplicate entries do not cause any problems since all the places where the frametable is currently used, the target operation happens to be idempotent. However, this is an undocumented fact, and any non-idempotent use of live_offsets from the frametable would lead to errors.

The fix is to simply sort and remove the duplicates as the number of live offsets is typically small.

alainfrisch commented:
What about making the type Emitaux.frame_descr private and expose a function `Emitaux.record_frame_descr`, which would take care of normalizing as suggested?  (Plus some ad hoc treatment for sparc which does not use this type.)

kayceesrk committed:
Remove duplicate live_offset entries from frametables

Reg.Set uses stamp for comparison. Hence, two virtual registers that map
to the same real register may have different stamps. This leads to
duplicate live_offset entries in the frametables. Fix is to simply sort
and remove the duplicates as the number of live offsets is small.

kayceesrk commented:
Thanks @alainfrisch for reviewing this PR. Sorry for the delay. 

I've updated the PR such that the type `frame_descr` and the list ref `frame_descriptors` are not exposed by `Emitaux`. Instead, the only function that is exposed is `record_frame_descr`. This function performs the deduplication and appends the frame descriptor to the list of frame descriptors. This way the mutable state is hidden behind the module interface, and the interface is cleaner as a result. 

As suggested, I've included the ad-hoc fix for sparc which does not use this type.

mshinwell commented:
This is OK, but I'm not sure if it should be marked as approved since KC hasn't yet signed a CLA.

@xavierleroy Please add the "approved" tag if this is not large enough for a CLA.  (I've asked KC to get in touch with you about this matter.)

kayceesrk commented:
Thanks @mshinwell.

xavierleroy commented:
I'm in touch with KC concerning the CLA.  The contribution is small enough that we can proceed without a CLA.
